frontend/src/components/CodeEditorV2.tsx

```diff
@@ -0,0 +1,130 @@
+import { EditorState, type Extension } from '@codemirror/state'
+import { EditorView, drawSelection, gutter, highlightActiveLineGutter, hoverTooltip, keymap, lineNumbers } from '@codemirror/view'
+
+import { bracketMatching, defaultHighlightStyle, foldGutter, foldKeymap, indentOnInput, syntaxHighlighting } from '@codemirror/language'
+import { lintGutter } from '@codemirror/lint'
+import { lintKeymap } from '@codemirror/lint'
+import { linter } from '@codemirror/lint'
+
+import { autocompletion, closeBrackets, closeBracketsKeymap } from '@codemirror/autocomplete'
+import { atomone } from '@uiw/codemirror-theme-atomone'
+import { githubLight } from '@uiw/codemirror-theme-github'
+
+import { defaultKeymap, history, indentWithTab, redo, undo } from '@codemirror/commands'
+import { handleRefresh, jsonSchemaHover, jsonSchemaLinter, stateExtensions } from 'codemirror-json-schema'
+import { json5, json5ParseLinter } from 'codemirror-json5'
+import { useEffect, useRef, useState } from 'react'
+import { useUserPreferences } from '../providers/user-preferences-provider'
+
+export const CodeEditor = ({
+  content,
+  schema,
+  readOnly,
+  onChange,
+  defaultContent,
+}: {
+  content: string
+  schema?: string
+  readOnly?: boolean
+  onChange?: (text: string) => void
+  defaultContent?: string
+}) => {
+  const { isDarkMode } = useUserPreferences()
+  const containerRef = useRef(null)
+  const viewRef = useRef<EditorView | null>(null)
+  const [lastReset, setLastReset] = useState(Date.now())
```

Comment:
Should this be optional? I'm guessing we'll still use this editor for displaying `json` and `schema` stuff that wouldn't need a `reset` button.

Reply:
Yes! The optionality is controlled by the `defaultContent` prop. I just have the `lastReset` state getting created no matter what because handling nullability on that gets really messy. I'll add a comment above that prop explaining it.